File: 322.coin-change.py
from typing import List


# @leet start
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        inf = 1_000_000
        # dp representing number of coins to reach here
        dp = [inf] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            for c in coins:
                if i - c >= 0:
                    dp[i] = min(dp[i], 1 + dp[i - c])
        if dp[amount] != inf:
            return dp[amount]
        else:
            return -1


# @leet end

File: 322.coin-change.java
import java.util.Arrays;

// @leet start
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        dp[0] = 0;
        Arrays.sort(coins);
        for (int i = 1; i <= amount; i++) {
            dp[i] = Integer.MAX_VALUE;
            for (int c : coins) {
                if (i - c < 0) {
                    continue;
                }
                if (dp[i - c] == Integer.MAX_VALUE) {
                    continue;
                }
                dp[i] = Math.min(dp[i - c] + 1, dp[i]);
            }
        }
        return dp[amount] == Integer.MAX_VALUE ? -1 : dp[amount];
    }
}
// @leet end
